### PR TITLE
Added graphiql-explorer to the GraphiQL page.

### PR DESCRIPTION
**Pull request by jrestall**
graphiql-explorer is a plugin for GraphiQL that adds a new technique to explore and build GraphQL queries. It adds a graphical representation of available fields and inputs that can be used in queries. It also allows constructing full queries by clicking through available fields and inputs without the repetitive process of typing these queries out by hand!

- Reduces the learning curve for creating GraphQL queries making the API more approachable to users.
- Helps users discover and use the more advanced fragment types that are necessary when using content items.
- Less typing! :tada: 

graphiql-explorer

This PR also updates graphiql and react dependencies. I'm using webpack to compile and pack the react & CSS code which is not something we've done before, so I'm expecting some comments on that. I inlined some CSS in Index.cshtml for simplicity and so that the spinner appears quickly but can refactor as needed.

### CODE REVIEW

**Comment by Skrypt** [comment]
I like the idea of introducing Webpack in the toolset. You removed the Assets.json file so I understand that it removes this module from the gulp pipeline? I need to test and I'll give you my review. I know that some of the packages here required to be built differently and we could not previously only use our Gulp pipeline because it requires to use Webpack. Maybe we could keep a hook with Gulp to trigger this webpack build. We'll see!

**Comment by jrestall** [comment]
Thanks @Skrypt! Yea this module no longer uses the gulp pipeline since webpack is doing all the bundling, along with compiling the React JSX and Typescript etc. A hook from the gulp pipeline sounds good to me though.

**Comment by Jetski5822** [comment]
In terms of what it does: Amazing!!! Seriously, this is a massive plus to our graphql offering.

Question: Does this now allow for a smoother upgrade path in regards to graphiql?

**Comment by jrestall** [comment]
Hey @Jetski5822, yea it's pretty sweet! Doesn't work with the orderBy and where filter clauses though, but good for quickly creating queries/mutations. Could do with some testing on a site with a lot of content types.

Upgrading graphiql should be as easy as increasing the version of 'graphiql' in the module's package.json and then running 'npm run-script build' and pushing the new graphiql-orchard.js and updated package.json file.

**Comment by Jetski5822** [comment]
Oh - how come it doesn't work for orderby or filters?

**Comment by carlwoodhouse** [comment]
looks pretty sweet @jrestall - cant wait to try it :)

**Comment by jrestall** [comment]
> Oh - how come it doesn't work for orderby or filters?

I'm not sure yet - It currently throws a JavaScript error when clicking either orderBy or where. It should hopefully be possible to get working.

> looks pretty sweet @jrestall - cant wait to try it :)

Thanks! Let me know if there's any issues besides the advanced filter criteria not working.

**Comment by jrestall** [comment]
> Oh - how come it doesn't work for orderby or filters?

Fixed 👏 

Screen Shot 2019-06-04 at 9 51 35 PM

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Assets/js/App.tsx [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+import GraphiQL from "graphiql";
+import GraphiQLExplorer from "graphiql-explorer";
+import { buildClientSchema, getIntrospectionQuery } from "graphql";
+
+import "graphiql/graphiql.css";
+import "../css/graphiql-orchard.css";
+
+function fetcher(params: Object): Promise<any> {
+    return fetch(
+        "/api/graphql",
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Assets/js/App.tsx]: It needs to be `@Href("~/api/graphql")`, so if it's in a script, you need to generate it in a well-known tag and fetch it from JS.

Comment (jrestall) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Assets/js/App.tsx]: Thanks @sebastienros, I've pushed the fix for this.

**Comment by sgrove** [comment]
@jrestall This looks great! Is there something we should change on our side related to the fix you had to do for `orderBy` and `where`?  Also, anything we could do (maybe more fragment support, advanced features, design, themability, etc.) that you would like to see for Orchard?  Also, I'd recommend for your power users hooking up this shortcut https://github.com/OneGraph/graphiql-explorer-example/blob/master/src/App.js#L104-L160 so that when there are lots of big operations in the document editor, you can shift-option/alt-click on a query and it'll scroll to it in explorer. [Here's an example document](https://serve.onegraph.com/short/DVBHMC) that shows how handy it can be (by the way, this comment was added from that document in the `AddOrchardCMSComment` mutation :D)

**Comment by jrestall** [comment]
Hi @sgrove! Firstly thank you for graphiql-explorer, it makes the experience of using graphql with Orchard Core so much better given how complex the content items can become. I understand you are working with the graphiql team to make it more seamless to integrate which would be great, though copying your example code to wire it in wasn't hard. 

The fix for the `orderBy `and `where` filters was related to my mistake, I referenced the 0.13.2 version of graphql initially since during build there were a ton of packages complaining about needing it as a peer dependency. After investigating the issue a bit it was clear I needed the latest version since some methods like `isRequiredInputField` only exist in the latest version. Maybe you could use npm to force a peer dependency on v14 since you are using isRequiredInputField which was only added recently https://github.com/graphql/graphql-js/pull/1465

One minor thing I've noticed is that when you add a lot of input fields, it doesn't wrap so you have to scroll horizontally to review what has been added. But each new input field that is added causes the scroll bar to reset to the left again which is mildly annoying. Wrapping to multiple lines to keep the query visible would be cool.

An example query with many arguments:
```
  menu(first: 10, orderBy: {publishedUtc: ASC}, where: {author_not_starts_with: "j"}, skip: 10) {
    contentItemVersionId
    contentType
    modifiedUtc
    published
  }
```

I'll look into adding that shortcut, but I don't think many users would easily discover that functionality so having a button like chrome dev tools inspect element crosshair could be better?

**Comment by mmdqq** [comment]
@jrestall jrestall I think add some code in OrchardCore.Apis.GraphQL.csproj ,than it can autocreate "graphiql-orchard.js" ,such as :  <Target Name="DebugRunWebpack" BeforeTargets="Build" Condition=" '$(Configuration)' == 'Debug' And !Exists('wwwroot\dist') ">
    
    <Exec Command="node --version" ContinueOnError="true">
      <Output TaskParameter="ExitCode" PropertyName="ErrorCode" />
    </Exec>
    <Error Condition="'$(ErrorCode)' != '0'" Text="Node.js is required to build and run this project. To continue, please install Node.js from https://nodejs.org/, and then restart your command prompt or IDE." />

    
    <Message Importance="high" Text="Performing first-run Webpack build..." />
    <Exec Command="node node_modules/webpack/bin/webpack.js" />
  </Target>
but i don't know webpack how to work in this project without "Microsoft.AspNetCore.SpaServices",can you help me?

**Comment by jrestall** [comment]
Hey @mmdqq, yea we could do that to generate graphiql-orchard.js on each build. It will slow the build and force everyone to have node installed as a prerequisite though so I'd be interested in input from yourself and @Skrypt, @sebastienros on the benefits over our current approach. 

The code would look like this:
```
<Target Name="RunWebpack" AfterTargets="ComputeFilesToPublish">
    <!-- As part of publishing, ensure the JS resources are freshly built in production mode -->
    <Exec Command="npm install" />
    <Exec Command="npm run-script build" />
  </Target>
```

**Comment by mmdqq** [comment]
@jrestall  Oh! Now I understand，but  I don't understand that how to Webpack  work with out "SpaServices"?

**Comment by jrestall** [comment]
SpaServices is an optional library that makes it easier to build single page applications on ASP.NET Core. It provides helpers for server-side prerendering, webpack dev middleware and routing.

We don't need any of that in this case, so using webpack directly is the simpler option.

**Comment by carlwoodhouse** [comment]
@jrestall @sgrove i just wanna say the explorer is badass :) especially with large schema's hah
image

**Comment by jrestall** [comment]
That's awesome to see @carlwoodhouse!

**Comment by valeryivan** [comment]
@jrestall This looks great! Is there something we should change on our side related to the fix you had to do for `orderBy` and `where`?
 
 Also, anything we could do (maybe more fragment support, advanced features, design, themability, etc.) that you would like to see for Orchard?
 
 Also, I'd recommend for your power users hooking up this shortcut https://github.com/OneGraph/graphiql-explorer-example/blob/master/src/App.js#L104-L160 so that when there are lots of big operations in the document editor, you can shift-option/alt-click on a query and it'll scroll to it in explorer. Here's an example document that shows how handy it can be (by the way, this comment was added from that document in the `AddOrchardCMSComment` mutation :D)

**Pull request merged**